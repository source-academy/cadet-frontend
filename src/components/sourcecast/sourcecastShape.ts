import { ExternalLibraryName } from '../assessment/assessmentShape';

export interface IInputTypeShape {
  chapterSelect: number;
  cursorPositionChange: IPosition;
  codeDelta: ICodeDelta;
  externalLibrarySelect: ExternalLibraryName;
  keyboardCommand: KeyboardCommand;
  selectionRangeData: ISelectionData;
}

export enum KeyboardCommand {
  run = 'run'
}

export enum PlaybackStatus {
  playing = 'playing',
  paused = 'paused'
}

export interface ICodeDelta {
  start: IPosition;
  end: IPosition;
  action: string;
  lines: string[];
}

export interface ISelectionRange {
  start: IPosition;
  end: IPosition;
}

export interface ISelectionData {
  range: ISelectionRange;
  isBackwards: boolean;
}

export interface IPosition {
  row: number;
  column: number;
}

// Refer: https://stackoverflow.com/questions/55758713/match-pair-for-keyof-and-valueof-an-interface
export type Input = keyof IInputTypeShape extends infer K
  ? K extends keyof IInputTypeShape
    ? { time: number; type: K; data: IInputTypeShape[K] }
    : never
  : never;

export interface IPlaybackData {
  init: {
    chapter: number;
    externalLibrary: ExternalLibraryName;
    editorValue: string;
  };
  inputs: Input[];
}

export interface ISourcecastData {
  title: string;
  description: string;
  inserted_at: string;
  updated_at: string;
  playbackData: string;
  id: number;
  uploader: {
    id: number;
    name: string;
  };
  url: string;
}

export enum RecordingStatus {
  notStarted = 'notStarted',
  recording = 'recording',
  paused = 'paused',
  finished = 'finished'
}
