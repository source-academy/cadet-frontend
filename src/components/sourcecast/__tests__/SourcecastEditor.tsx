import * as React from 'react';

import { shallow } from 'enzyme';

import SourcecastEditor, { ISourcecastEditorProps } from '../SourcecastEditor';
import { IPosition } from '../sourcecastShape';

const componentDidMountSpy = jest.fn();

jest
  .spyOn(SourcecastEditor.prototype, 'componentDidMount')
  .mockImplementation(componentDidMountSpy);

test('SourcecastEditor renders correctly', () => {
  const props: ISourcecastEditorProps = {
    breakpoints: [],
    codeDeltasToApply: [],
    editorReadonly: false,
    editorSessionId: '',
    editorValue: '',
    highlightedLines: [],
    isEditorAutorun: false,
    inputToApply: null,
    isPlaying: false,
    isRecording: false,
    sharedbAceInitValue: '',
    sharedbAceIsInviting: false,
    getTimerDuration: () => 1,
    handleDeclarationNavigate: (cursorPosition: IPosition) => {},
    handleEditorEval: () => {},
    handleEditorValueChange: newCode => {},
    handleEditorUpdateBreakpoints: breakpoints => {},
    handleFinishInvite: () => {},
    handleRecordInput: input => {},
    handleSetWebsocketStatus: websocketStatus => {},
    handleUpdateHasUnsavedChanges: hasUnsavedChanges => {}
  };
  const app = <SourcecastEditor {...props} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
  expect(componentDidMountSpy).toHaveBeenCalled();
});
