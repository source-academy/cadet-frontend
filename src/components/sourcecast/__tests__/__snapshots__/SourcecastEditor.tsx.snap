// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourcecastEditor renders correctly 1`] = `
"<HotKeys className=\\"Editor\\" handlers={{...}}>
  <div className=\\"row editor-react-ace\\">
    <ReactAce className=\\"react-ace\\" commands={{...}} editorProps={{...}} markers={{...}} fontSize={17} height=\\"100%\\" highlightActiveLine={false} mode=\\"javascript\\" onChange={[Function]} onCursorChange={[Function]} onSelectionChange={[Function]} readOnly={false} theme=\\"source\\" value=\\"\\" width=\\"100%\\" setOptions={{...}} name=\\"ace-editor\\" focus={false} enableSnippets={false} showGutter={true} onPaste={{...}} onLoad={{...}} onScroll={{...}} minLines={{...}} maxLines={{...}} showPrintMargin={true} tabSize={4} cursorStart={1} style={{...}} scrollMargin={{...}} wrapEnabled={false} enableBasicAutocompletion={false} enableLiveAutocompletion={false} placeholder={{...}} navigateToFileEnd={true} />
  </div>
</HotKeys>"
`;
